pkg/controllers/nodeclass/garbagecollection/controller.go

```diff
@@ -0,0 +1,155 @@
+	return &Controller{
+		kubeClient:              kubeClient,
+		cloudProvider:           cloudProvider,
+		instanceProfileProvider: instanceProfileProvider,
+		region:                  region,
+	}
+}
+
+func (c *Controller) getActiveProfiles(ctx context.Context) (sets.Set[string], error) {
+	nodeClaims, err := nodeclaimutils.ListManaged(ctx, c.kubeClient, c.cloudProvider)
+	if err != nil {
+		return nil, fmt.Errorf("listing nodeclaims, %w", err)
+	}
+
+	activeProfiles := sets.New[string]()
+	for _, nc := range nodeClaims {
+		if profileName, ok := nc.Annotations[v1.AnnotationInstanceProfile]; ok {
+			activeProfiles.Insert(profileName)
+		}
+
+	}
+
+	return activeProfiles, nil
+}
+
+func (c *Controller) getCurrentProfiles(ctx context.Context) (sets.Set[string], error) {
+	nodeClasses := &v1.EC2NodeClassList{}
+	if err := c.kubeClient.List(ctx, nodeClasses); err != nil {
+		return nil, fmt.Errorf("listing nodeclasses, %w", err)
+	}
+
+	currentProfiles := sets.New[string]()
+
+	for _, nc := range nodeClasses.Items {
+		if nc.Status.InstanceProfile != "" {
+			currentProfiles.Insert(nc.Status.InstanceProfile)
+		}
+	}
+	return currentProfiles, nil
+}
+
+func (c *Controller) shouldDeleteProfile(profileName string, currentProfiles sets.Set[string], activeProfiles sets.Set[string]) bool {
+	// Skip if this is a current profile in any EC2NodeClass
+	if _, ok := currentProfiles[profileName]; ok {
+		return false
+	}
+
+	// Skip if this is a protected profile
+	if c.instanceProfileProvider.IsProtected(profileName) {
+		return false
+	}
+
+	// Skip if this is an active profile (NodeClaims are using it)
+	if _, isActive := activeProfiles[profileName]; isActive {
+		return false
+	}
+
+	return true
+}
+
+func (c *Controller) cleanupInactiveProfiles(ctx context.Context, activeProfiles sets.Set[string], currentProfiles sets.Set[string]) error {
+	profiles, err := c.instanceProfileProvider.ListClusterProfiles(ctx)
+
+	if err != nil {
+		return fmt.Errorf("listing instance profiles, %w", err)
+	}
+
+	for _, profile := range profiles {
+		profileName := *profile.InstanceProfileName
+
+		if !c.shouldDeleteProfile(profileName, currentProfiles, activeProfiles) {
+			continue
+		}
+
+		if err := c.instanceProfileProvider.Delete(ctx, profileName); err != nil {
+			return serrors.Wrap(fmt.Errorf("deleting instance profile, %w", err), "instance-profile", profileName)
+		}
+		log.FromContext(ctx).V(1).Info("deleted instance profile", "instance-profile", profileName)
+	}
+	return nil
+}
+
+func (c *Controller) Reconcile(ctx context.Context) (reconcile.Result, error) {
+	ctx = log.IntoContext(ctx, log.FromContext(ctx).WithValues("controller", "instanceprofile.garbagecollection"))
+
+	activeProfiles, err := c.getActiveProfiles(ctx)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	currentProfiles, err := c.getCurrentProfiles(ctx)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
```

Comment:
nit: I'd consider resolving these in `cleanupInstanceProfiles` directly since they're not used elsewhere.